datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java

```diff
@@ -1,173 +1,291 @@
-                    MDCBuilder.buildMdcContext(sdxCluster);
-                    try {
-                        transactionService.required(() -> {
-                            SdxCluster created = sdxClusterRepository.save(sdxCluster);
-                            ownerAssignmentService.assignResourceOwnerRoleIfEntitled(created.getInitiatorUserCrn(), created.getCrn(), created.getAccountId());
-                            sdxStatusService.setStatusForDatalake(DatalakeStatusEnum.REQUESTED, "Data lake requested", created);
-                            context.setSdxId(created.getId());
-                            return created;
-                        });
-                        LOGGER.info("Detaching Sdx with Name: {} and Crn: {} is complete", sdxCluster.getName(), sdxCluster.getCrn());
-                    } catch (TransactionService.TransactionExecutionException e) {
-                        throw new TransactionService.TransactionRuntimeExecutionException(e);
-                    }
-                }
-                sendEvent(context, SDX_DETACH_FINALIZED_EVENT.event(), payload);
+            protected void doExecute(SdxContext context, SdxEvent payload,
+                    Map<Object, Object> variables) throws Exception {
+                SdxCluster detachedCluster = (SdxCluster) variables.get(DETACHED_SDX);
+                LOGGER.info("Detaching of SDX with ID {} finished.", detachedCluster.getId());
+
+                sdxStatusService.setStatusForDatalakeAndNotify(
+                        DatalakeStatusEnum.STOPPED, "Datalake is detached.", detachedCluster.getId()
+                );
+
+                SdxCluster resizedCluster = (SdxCluster) variables.get(RESIZED_SDX);
+                LOGGER.info("Attaching of SDX cluster with ID {} in progress.", resizedCluster.getId());
+
+                MDCBuilder.buildMdcContext(resizedCluster);
+                resizedCluster = sdxAttachService.saveSdxAndAssignResourceOwnerRole(resizedCluster);
+                sdxStatusService.setStatusForDatalake(
+                        DatalakeStatusEnum.REQUESTED, "Newly attached datalake requested.", resizedCluster
+                );
+
+                LOGGER.info("Attaching of SDX cluster with ID {} is complete.", resizedCluster.getId());
+                context.setSdxId(resizedCluster.getId());
+                sendEvent(context, SDX_ATTACH_NEW_CLUSTER_SUCCESS_EVENT.event(), payload);
             }
 
             @Override
-            protected Object getFailurePayload(SdxDetachSuccessEvent payload, Optional<SdxContext> flowContext, Exception ex) {
-                return null;
+            protected Object getFailurePayload(SdxEvent payload, Optional<SdxContext> flowContext,
+                    Exception e) {
+                return SdxDetachFailedEvent.from(payload, e);
+            }
+        };
+    }
+
+    @Bean(name = "SDX_ATTACH_NEW_CLUSTER_FAILED_STATE")
+    public Action<?, ?> sdxAttachNewClusterFailedAction() {
```

Comment:
I'm also a fan of having this action and being able to have the flow get back into the initial state with automatic recovery.
If this is successful, and we see a retry attempted we might need some logic similar to how we use the FULLY_RESTARTABLE_FLOWS in SDXRetryService to make sure we are starting off from the correct place. Or just make it clear that we aren't retrying from this point.